x/ccv/child/module.go

```diff
@@ -263,25 +272,71 @@ func (am AppModule) OnChanOpenTry(
 	counterparty channeltypes.Counterparty,
 	version,
 	counterpartyVersion string,
-) error {
-	return sdkerrors.Wrap(ccv.ErrInvalidChannelFlow, "channel handshake must be initiated by child chain")
+) (string, error) {
+	return "", sdkerrors.Wrap(
+		ccv.ErrInvalidChannelFlow, "channel handshake must be initiated by child chain")
 }
 
 // OnChanOpenAck implements the IBCModule interface
 func (am AppModule) OnChanOpenAck(
 	ctx sdk.Context,
 	portID,
 	channelID string,
-	counterpartyVersion string,
+	counterpartyMetadata string,
 ) error {
 	// ensure parent channel has already been created
 	if parentChannel, ok := am.keeper.GetParentChannel(ctx); ok {
-		return sdkerrors.Wrapf(ccv.ErrDuplicateChannel, "parent channel: %s already established", parentChannel)
+		return sdkerrors.Wrapf(ccv.ErrDuplicateChannel,
+			"parent channel: %s already established", parentChannel)
+	}
+
+	var md parenttypes.HandshakeMetadata
+	if err := (&md).Unmarshal([]byte(counterpartyMetadata)); err != nil {
+		return sdkerrors.Wrapf(ccv.ErrInvalidHandshakeMetadata,
+			"error unmarshalling ibc-ack metadata: ", err)
+	}
+
+	if md.Version != ccv.Version {
+		return sdkerrors.Wrapf(ccv.ErrInvalidVersion,
+			"invalid counterparty version: %s, expected %s", md.Version, ccv.Version)
 	}
 
-	if counterpartyVersion != ccv.Version {
-		return sdkerrors.Wrapf(ccv.ErrInvalidVersion, "invalid counterparty version: %s, expected %s", counterpartyVersion, ccv.Version)
+	am.keeper.SetProviderFeePoolAddrStr(ctx, md.ProviderFeePoolAddr)
+
+	///////////////////////////////////////////////////
+	// Initialize distribution token transfer channel
+	// NOTE the handshake for this channel is handled by
+	//      the ibc-go/transfer module
+
+	// here NewCounterparty asks for a counterparty channel with the port
+	// transfer. ChannelID (second input) is left blank, since it will be
+	// autogenerated by the counterparty and returned to us when calling
+	// ChannelOpenInit
+	transferCounterparty := channeltypes.NewCounterparty(transfertypes.Version, "")
+
+	distChan := channeltypes.Channel{
+		State:        channeltypes.UNINITIALIZED,
+		Ordering:     channeltypes.UNORDERED,
+		Counterparty: transferCounterparty,
+
+		// XXX can either pass back from provider OnChanOpenTry through
+		// version-metadata (easy) or store on the consumer chain in
+		// OnChanOpenInit (little bit more difficult but also do-able)
+		ConnectionHops: "", //connectionHops,
+
+		Version: transfertypes.Version,
 	}
+	distrTransferMsg := channeltypes.MsgChannelOpenInit{
```

Comment:
This is incorrect. Use the constructor function: https://github.com/cosmos/ibc-go/blob/main/modules/core/04-channel/types/msgs.go#L19

ConnectionHops should not be empty. Use the connection hops that this channel has.

Get connectionhops here by:

```go
// pseudocode
channel := am.keeper.GetChannel(ctx, portID, channelID)
connectionHops := channel.ConnectionHops
```

Reply:
Thank you. I've updated the code to using the constructor function (didn't know it was there). Let me know what you think, I think it's correct now